lib/server.js

```diff
@@ -1143,19 +1143,37 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
     // the connection (or we lose the connection).
     // we consider a closed request as flushed from metrics point of view
     function onClose() {
-        res._flushed = true;
-        req._timeFlushed = process.hrtime();
-
         res.removeListener('finish', onFinish);
         res.removeListener('error', onError);
+
         req._connectionState = 'close';
 
+        // Do no handle close if request is already flushed
+        // In Node >= 10 res's close event is always emitted
+        // Introduced in: https://github.com/nodejs/node/pull/20611
+        if (res._flushed) {
+            return;
+        }
+
+        // Handle close and return error if request is not flushed yet
+        res._flushed = true;
+        req._timeFlushed = process.hrtime();
+
         // Request was aborted or closed. Override the status code
         res.statusCode = 444;
-
-        self._finishReqResCycle(req, res, new errors.RequestCloseError());
+        self._finishReqResCycle(
+            req,
+            res,
+            res.err || new errors.RequestCloseError()
+        );
     }
-    req.once('close', onClose);
+
+    // Handle res's close on to the next cycle
+    // Required in Node >= 10:
+    // introduced in: https://github.com/nodejs/node/pull/20941
+    req.once('close', function onReqClose() {
```

Comment:
Is it that because _res_' `'close'` event can now be emitted using `process.nextTick` we want to handle _req_'s `'close'` event on the actual next tick of the libuv event loop so that we process req's `'close'` event _after_ `res`'s `'close'` event?

It seems this ordering could break if .e.g node started emitting the res' `'close'` event using `setImmediate`, which seems a bit brittle.